File: src-tauri/src/ocr.rs
// Calls the Anthropic Messages API (vision) to OCR a math equation image
// into raw LaTeX source. The LaTeX -> Typst conversion itself happens on the
// frontend (src/lib/converter.js), which is already unit-tested there; this
// module's only job is "image in, LaTeX text out".

use serde::{Deserialize, Serialize};
use std::time::Duration;

const ANTHROPIC_API_URL: &str = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_VERSION: &str = "2023-06-01";

const SYSTEM_PROMPT: &str = "You are an OCR engine specialised in mathematical notation. \
Transcribe the mathematical expression(s) in the image into LaTeX source. \
Output ONLY the raw LaTeX - no markdown code fences, no $ or $$ delimiters, no \\[ \\], \
no explanation, and no surrounding prose. \
Preserve structure exactly: use align*/aligned, cases, pmatrix/bmatrix/vmatrix, etc. \
where the image shows multi-line or matrix layouts. If the image contains no \
recognisable math, output exactly: NO_EQUATION_FOUND";

#[derive(Serialize)]
struct ImageSource<'a> {
    #[serde(rename = "type")]
    kind: &'a str,
    media_type: &'a str,
    data: String,
}

#[derive(Serialize)]
#[serde(tag = "type", rename_all = "lowercase")]
enum ContentBlock<'a> {
    Image { source: ImageSource<'a> },
    Text { text: String },
}

#[derive(Serialize)]
struct Message<'a> {
    role: &'a str,
    content: Vec<ContentBlock<'a>>,
}

#[derive(Serialize)]
struct MessagesRequest<'a> {
    model: &'a str,
    max_tokens: u32,
    system: &'a str,
    messages: Vec<Message<'a>>,
}

#[derive(Deserialize)]
struct ResponseTextBlock {
    #[serde(rename = "type")]
    kind: String,
    text: Option<String>,
}

#[derive(Deserialize)]
struct MessagesResponse {
    #[serde(default)]
    content: Vec<ResponseTextBlock>,
    #[serde(default)]
    error: Option<ApiError>,
}

#[derive(Deserialize)]
struct ApiError {
    message: String,
}

/// Sends a base64-encoded PNG to Claude and returns the raw LaTeX it
/// transcribed. Returns `Err` with a user-displayable message on any
/// network, auth, or parsing failure.
pub async fn recognize_latex(
    api_key: &str,
    model: &str,
    png_base64: String,
) -> Result<String, String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(45))
        .build()
        .map_err(|e| format!("could not build HTTP client: {e}"))?;

    let body = MessagesRequest {
        model,
        max_tokens: 1024,
        system: SYSTEM_PROMPT,
        messages: vec![Message {
            role: "user",
            content: vec![
                ContentBlock::Image {
                    source: ImageSource {
                        kind: "base64",
                        media_type: "image/png",
                        data: png_base64,
                    },
                },
                ContentBlock::Text {
                    text: "Transcribe the equation in this image to LaTeX.".to_string(),
                },
            ],
        }],
    };

    let response = client
        .post(ANTHROPIC_API_URL)
        .header("x-api-key", api_key)
        .header("anthropic-version", ANTHROPIC_VERSION)
        .header("content-type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("could not reach Anthropic API: {e}"))?;

    let status = response.status();
    let parsed: MessagesResponse = response
        .json()
        .await
        .map_err(|e| format!("could not parse Anthropic API response: {e}"))?;

    if let Some(err) = parsed.error {
        return Err(format!("Anthropic API error: {}", err.message));
    }
    if !status.is_success() {
        return Err(format!("Anthropic API returned HTTP {status}"));
    }

    let raw = parsed
        .content
        .into_iter()
        .find(|b| b.kind == "text")
        .and_then(|b| b.text)
        .ok_or_else(|| "Anthropic API response had no text content".to_string())?;

    let cleaned = strip_code_fence(raw.trim());

    if cleaned == "NO_EQUATION_FOUND" || cleaned.is_empty() {
        return Err("No equation was found in the selected region.".to_string());
    }

    Ok(cleaned.to_string())
}

/// Defensively strips a ```latex ... ``` / ``` ... ``` wrapper in case the
/// model adds one despite being told not to.
fn strip_code_fence(s: &str) -> &str {
    let s = s.trim();
    if let Some(stripped) = s.strip_prefix("```") {
        let stripped = stripped.trim_start_matches("latex").trim_start_matches("tex");
        let stripped = stripped.trim_start_matches('\n');
        if let Some(end) = stripped.rfind("```") {
            return stripped[..end].trim();
        }
        return stripped.trim();
    }
    s
}

File: web_app/api/ocr.js
// Vercel serverless function: proxies a single vision-OCR request to
// Anthropic's Messages API.
//
// Why this exists at all (rather than calling Anthropic directly from the
// browser): Anthropic's API does not allow plain cross-origin browser
// requests, and even if it did, putting the user's API key directly in a
// browser-originated request is the kind of thing that ages badly. So the
// browser calls this same-origin endpoint instead, which forwards the
// request server-side.
//
// API-key model: bring-your-own-key, same as the desktop app. The key is
// supplied by the client on every request body and is never stored,
// logged, or cached here - it only ever exists in memory for the duration
// of this one request. There is no Anthropic key configured as a Vercel
// environment variable, and none is required to deploy this project.
//
// Mirrors the prompt/parsing logic of the desktop app's
// src-tauri/src/ocr.rs 1:1 so OCR behaviour is identical between the two.

export const config = {
  api: {
    bodyParser: {
      // Base64-encoded PNGs inflate ~37% over raw bytes; this leaves enough
      // room for a reasonably large snip/upload while staying under
      // Vercel's platform-level request body ceiling.
      sizeLimit: "8mb",
    },
  },
};

const ANTHROPIC_API_URL = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_VERSION = "2023-06-01";
const DEFAULT_MODEL = "claude-sonnet-4-6";
const REQUEST_TIMEOUT_MS = 45_000;

const SYSTEM_PROMPT = `You are an OCR engine specialised in mathematical notation. \
Transcribe the mathematical expression(s) in the image into LaTeX source. \
Output ONLY the raw LaTeX - no markdown code fences, no $ or $$ delimiters, no \\[ \\], \
no explanation, and no surrounding prose. \
Preserve structure exactly: use align*/aligned, cases, pmatrix/bmatrix/vmatrix, etc. \
where the image shows multi-line or matrix layouts. If the image contains no \
recognisable math, output exactly: NO_EQUATION_FOUND`;

/** Defensively strips a ```latex ... ``` / ``` ... ``` wrapper in case the
 * model adds one despite being told not to. Same behaviour as the desktop
 * app's strip_code_fence in src-tauri/src/ocr.rs. */
function stripCodeFence(raw) {
  const s = raw.trim();
  if (!s.startsWith("```")) return s;
  let stripped = s.slice(3);
  stripped = stripped.replace(/^(latex|tex)/, "");
  stripped = stripped.replace(/^\n/, "");
  const end = stripped.lastIndexOf("```");
  return (end !== -1 ? stripped.slice(0, end) : stripped).trim();
}

export default async function handler(req, res) {
  if (req.method !== "POST") {
    res.setHeader("Allow", "POST");
    return res.status(405).json({ error: "Method not allowed." });
  }

  const { apiKey, model, imageBase64, mediaType } = req.body || {};

  if (!apiKey || typeof apiKey !== "string" || !apiKey.trim()) {
    return res.status(400).json({
      error: "Missing Anthropic API key. Add one in Settings before snipping or uploading.",
    });
  }
  if (!imageBase64 || typeof imageBase64 !== "string") {
    return res.status(400).json({ error: "Missing image data." });
  }

  const usedModel = (model && String(model).trim()) || DEFAULT_MODEL;
  const usedMediaType = mediaType && String(mediaType).startsWith("image/") ? mediaType : "image/png";

  const anthropicBody = {
    model: usedModel,
    max_tokens: 1024,
    system: SYSTEM_PROMPT,
    messages: [
      {
        role: "user",
        content: [
          {
            type: "image",
            source: {
              type: "base64",
              media_type: usedMediaType,
              data: imageBase64,
            },
          },
          { type: "text", text: "Transcribe the equation in this image to LaTeX." },
        ],
      },
    ],
  };

  let response;
  try {
    response = await fetch(ANTHROPIC_API_URL, {
      method: "POST",
      headers: {
        "x-api-key": apiKey.trim(),
        "anthropic-version": ANTHROPIC_VERSION,
        "content-type": "application/json",
      },
      body: JSON.stringify(anthropicBody),
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
  } catch (err) {
    return res.status(502).json({ error: `Could not reach Anthropic API: ${err.message || err}` });
  }

  let parsed;
  try {
    parsed = await response.json();
  } catch {
    return res.status(502).json({ error: "Could not parse Anthropic API response." });
  }

  if (parsed && parsed.error) {
    const status = response.status >= 400 ? response.status : 502;
    return res.status(status).json({ error: `Anthropic API error: ${parsed.error.message || JSON.stringify(parsed.error)}` });
  }
  if (!response.ok) {
    return res.status(response.status).json({ error: `Anthropic API returned HTTP ${response.status}` });
  }

  const textBlock = Array.isArray(parsed.content) ? parsed.content.find((b) => b.type === "text") : null;
  const raw = textBlock && textBlock.text;
  if (!raw) {
    return res.status(502).json({ error: "Anthropic API response had no text content." });
  }

  const cleaned = stripCodeFence(raw.trim());
  if (cleaned === "NO_EQUATION_FOUND" || cleaned.length === 0) {
    return res.status(422).json({ error: "No equation was found in the selected region." });
  }

  return res.status(200).json({ latex: cleaned });
}
